# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [5]:
def addTwo(x):
    return x + 2

print(addTwo(4))

#Lambda version...

a_num = 4
print((lambda x: x + 2)(a_num))

6
6


#### Saving to a Variable

In [6]:
f_test = lambda x: x - 2

f_test(10)

8

#### Multiple Inputs

In [8]:
print((lambda x, y, z: x * y)(3, 6, 9))

x_test = lambda x, y, z: x * y

print(x_test(3,5,8))

18
15


#### Passing a Lambda into a Function

In [9]:
def multiply(f, num):
    
    return f(num)

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [12]:
def multiply_tst(num):
    return num * 4

def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

def returnLam(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x105b08310>
8
19


#### If Statements within Lambdas

In [13]:
f_condition = lambda num : num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))
print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
# Creating an anonymous function (lambda) to cube the input
f = lambda x: x ** 3

# Getting input from the user
user_input = float(input("Enter a number: "))

# Using the anonymous function to cube the input
result = f(user_input)

# Displaying the result
print(f"The cube of {user_input} is: {result}")


Enter a number: 3
The cube of 3.0 is: 27.0


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [17]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 **2
    else:
        return num, num2
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_map = list(map(squared, numbers, more_nums))
print(squared_nums_map)

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [18]:
print(list(map(lambda x, y: (x ** 2, y ** 3) if x < 10 and y < 10 else (x,y), numbers, more_nums)))

[(16, 64), (11, 10), (20, 3), (9, 8), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [19]:
# Get a list of numbers from the user
user_input = input("Enter a list of numbers separated by space: ")
numbers = list(map(float, user_input.split()))

# Use map and a lambda function to double each number and subtract one
result = list(map(lambda x: 2*x - 1, numbers))

# Displaying the result
print("Original numbers:", numbers)
print("Transformed numbers:", result)


Enter a list of numbers separated by space: 6 7 9
Original numbers: [6.0, 7.0, 9.0]
Transformed numbers: [11.0, 13.0, 17.0]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [1]:
names = ['bob', 'andy', 'gerald', 'david', 'AngeLica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    elif name[0].lower() == "b":
        return True
    else:
        return False

new_names_list = list(filter(a_names, names))
print(new_names_list)


['bob', 'andy', 'AngeLica']


#### Using Lambda's with Filter()

In [2]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lambda)

['andy', 'AngeLica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [3]:
# Get a list of numbers from the user
user_input = input("Enter a list of numbers separated by space: ")
numbers = list(map(float, user_input.split()))

# Calculate the mean of the list
mean = sum(numbers) / len(numbers)

# Use filter and a lambda function to filter out numbers below the mean
filtered_numbers = list(filter(lambda x: x >= mean, numbers))

# Displaying the result
print("Original numbers:", numbers)
print("Numbers above or equal to the mean:", filtered_numbers)


Enter a list of numbers separated by space: 3 4 5 6 7
Original numbers: [3.0, 4.0, 5.0, 6.0, 7.0]
Numbers above or equal to the mean: [5.0, 6.0, 7.0]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [9]:
from functools import reduce

list_1 = [2, 4, 6, 8 , 10]

def addNums (num1, num2):
    return num1 + num2

result_add = reduce(addNums, list_1)

print(result_add)

#subtract list of numbers

def subtractNums (num1, num2):
    return num1 - num2
result_sub = reduce(subtractNums, list_1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [13]:
result_lamb = reduce(lambda x, y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [14]:
# Example list of numbers
numbers = [2, 3, 4, 5]

# Use reduce and a lambda function to multiply the numbers together
result = reduce(lambda x, y: x * y, numbers)

# Displaying the result
print("Original numbers:", numbers)
print("Product of the numbers:", result)

Original numbers: [2, 3, 4, 5]
Product of the numbers: 120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [25]:
def addNums(num):
#Here is our base case...
    if num <= 1:
        print('addNums(1) = 1')
        return num
    else:
        print(f'addNums({num}) = {num -1}')
        return num + addNums(num - 1)
    
addNums(5)
    

addNums(5) = 4
addNums(4) = 3
addNums(3) = 2
addNums(2) = 1
addNums(1) = 1


15

#### Writing a Factorial Function

In [26]:
def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)

factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [27]:
def recursive_subtraction(n):
    # Base case: if n is 0, return 0
    if n == 0:
        return 0
    # Recursive case: subtract n from the result of the function with n-1
    else:
        return n - recursive_subtraction(n - 1)

# Example usage
argument = 5
result = recursive_subtraction(argument)

# Displaying the result
print(f"The result of subtracting all numbers down to {argument} is: {result}")

The result of subtracting all numbers down to 5 is: 3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [31]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start
        start += step

        for i in my_range(20, start = 2):
            my_generator_value = i
            print(my_generator_value)

my_range(20, start = 2)

<generator object my_range at 0x10693ec00>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [34]:
# Example usage
number_argument = 4

# Using the generator and printing each squared number until zero is reached
for i in reversed(range(number_argument + 1)):
    squared_number = i ** 2
    print(f"{i} squared is: {squared_number}")


4 squared is: 16
3 squared is: 9
2 squared is: 4
1 squared is: 1
0 squared is: 0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [35]:
# Example list with empty strings
places = ["", "Argentina", "", "San Diego", "Boston", "", "New York", ""]

# Use filter and a lambda function to filter out empty strings
filtered_places = list(filter(lambda x: x != "", places))

# Displaying the result
print(filtered_places)


places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [36]:
# Example list with names
names = ['David hassELHOFF', 'Gary A.J. Bernstein', 'Andrew P. Garfield', 'Joel Carter', 'Victor aNisimov']

# Use sorted with a lambda function as the key to sort by last name
sorted_names = sorted(names, key=lambda x: x.split()[-1].capitalize())

# Displaying the result
print(sorted_names)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [37]:
# Example list of tuples with city and temperature in Celsius
temperature_data = [('Nashua', 32), ('Boston', 12), ('Los Angeles', 44), ('Miami', 29)]

# Use map and a lambda function to convert Celsius to Fahrenheit for each tuple
converted_temperatures = list(map(lambda data: (data[0], 9/5 * data[1] + 32), temperature_data))

# Displaying the result
print(converted_temperatures)



[('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [40]:
def fibonacci(n, a=0, b=1, current_iteration=0):
    if current_iteration > n:
        return
    print(f"Iteration {current_iteration}: {b}")
    fibonacci(n, b, a + b, current_iteration + 1)

# Example usage
fibonacci(5)


Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
